---
title: "COO PCA scRNAseq"
author: "Onur BASAK"
date: "9/3/2022"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    number_sections: false
    theme: united
    highlight: tango
    css: COO.css
---

```{r setup, include=FALSE}
library(knitr)
knitr::opts_chunk$set(echo = TRUE)
```

# Preperation

## Install packages
You need to do this only once! The script below checks whether the listed CRAN packages are installed on your computer. If they are not, they will be automatically installed.

```{r install_CRAN_packages, include=FALSE, eval=FALSE}
# Install some of the required packaged from CRAN
## First specify the packages of interest
packages = c("devtools","hdf5r","dplyr","ggplot2","stringr",
             "RColorBrewer","useful","readr","BiocManager","factoextra")

## Now load or install&load all
package.check <- lapply(
  packages,
  FUN = function(x) {
    if (!require(x, character.only = TRUE)) {
      install.packages(x, dependencies = TRUE)
      library(x, character.only = TRUE)
    }
  }
)

```

Seurat for scRNAseq data analysis
```{r install_Seurat, include=FALSE, eval=FALSE}
install.packages('Seurat')
```

The following is necessary for interaction with hdf5 objects, and can only be installed using BiocManager installer
```{r, install_rhdf5, include=FALSE, eval=FALSE}
BiocManager::install("rhdf5")
```

## Set working directory
You may need to run this on the terminal, since in some environment the working directory is set to ~ after the chunk is run! An alternative is to use the pull down at the top of the bottom-right window on RStudio click Files/More to see the option
```{r setwd}
setwd("/Users/onurbasak/Documents/1_BCRM/11 teaching/lectures/Bachelor/Bioinformatics_BMW_firstyear/2022/COO")
```

# 1. Introduction
In this COO you will learn how to analyse single cell RNA sequencing data (scRNAseq). For this, we will use the well established Seurat pipeline. This pipeline include function that do most of the work 'behind the scenes'. This makes the tool very user friendly and suitable for todays tutorial. There is extensive documentation on the use of the pipeline. The following tutorial is the closest to what we will do today: <https://satijalab.org/seurat/articles/pbmc3k_tutorial.html>

**Content summary**
*1)* principle component analysis (PCA) using prcomp()
Even though dimensionality reduction methods are already integrated in the pipeline, I would like to gain a bit more insight into how the actual code for on of the parts of the analysis looks like. Thus, we will start by performing PCA on single cell data using the prcomp() function, which is part of the 'stats' package that is an integral component of R.

*2)* scRNAseq data analysis using Seurat

# 2. Source data
For today's tutorial, we will use the scRNAseq atlas of the adolescent mouse cortex published by Saunders et al 2018. This data is extensive and is available at mousebrain.org, which we briefly discussed at the end of hte lecture. There is an online tool with which you can browse the data as well.

## 2.1 download the data
The data downloaded and processed into a 'Seurat object' to prevent technical errors caused by the 'loom' file format in several computers. You can download it from the blackboard, the in COO/data folder, or from the Following Github page made for this course: <>
'Linnerson_cortex_10X22.rds'

### Load libraries

```{r load_libraries, echo=FALSE,verbose = FALSE}
library(dplyr,verbose = FALSE)
library(ggplot2,verbose = FALSE)
library(stringr,verbose = FALSE)
library(Seurat,verbose = FALSE)
library(RColorBrewer,verbose = FALSE)
library(useful,verbose = FALSE)
library(readr,verbose = FALSE)
library(hdf5r,verbose = FALSE)
library(factoextra,verbose = FALSE)
```


### 2.2 Load the Seurat object

```{r load_data}
adult_cortex <- readRDS(file = 'Linnerson_cortex_10X22.rds')
adult_cortex
```

***Note:*** The object contains data from 6658 cells (samples) and 27998 features (genes). There is 1 assay (RNA)

### 2.3 Shape the dataframe
We will start by taking the data matrix out and generating a dataframe for use in PCS analysis

```{r shape_dataframe}
cortex_df = as.data.frame(adult_cortex@assays$RNA@counts)
kable(head(cortex_df[,1:3]),digits = 6)
```
You can see the gene names as rownames and cell names as column names. Cell names contain the source 10x run (3 different runs e,g, cortex1), sample in each run (_1 or _2) and the cell barcode that was used for demultiplexing. This information wont really be useful for this practical.
The numbers indicate the number of reads, or in another words, mRNA molecules detected for hte gene in hte respective cell. This is after correcting for PCR duplicates using the unique molecular identifier (UMI). This process was done beforehand, during hte alignment step.

### 2.4 downsize the dataset
To save from time, we will subset the data, randomly. We will do this by selecting 666 random cells (10% of the data). Change the number is you are superficial

```{r downsize}
cortex_df_sub = cortex_df[,sample(ncol(cortex_df), 666)]
kable(head(cortex_df_sub[,1:3]),digits = 6)
```
***Note:*** There are a lot of zeros!
object contains data from 6658 cells (samples) and 27998 features (genes). There is 1 assay (RNA)

### 2.5 invert the dataset 
We need to invert the dataset for the PCA analysis. Why? The package will compare the differences between rows and we want to analyse the differences between cells. By inverting hte dataset, we will place cells into row.
If you omit this, you will be analysing the variation between genes! If this is your question,of course, you can do that.

```{r invert}
cortex_df_sub_t = as.data.frame(t(cortex_df_sub))
kable(head(cortex_df_sub_t[,1:3]),digits = 6)
```
### 2.5 scale the dataset 
Standardization is simply calculating the standard deviation from the mean. The mean becomes zero, a value of +1 means the gene is expressed 1 st. dev. higher than the average.

Why do we do this? If we dont, we will be mostly comparing the differences in absolute expression values of genes, This means highly expressed genes will dominate the analysis. 

However, is we scale (standardise), we will be looking at the pattern. We can now compare different genes independent of their expression level.

```{r scale}
cortex_df_sub_t_scale = scale(cortex_df_sub_t) #scale 
cortex_df_sub_t_scale = data.frame(cortex_df_sub_t_scale) #convert to a dataframe
cortex_df_sub_t_scale[is.na(cortex_df_sub_t_scale)] = 0 #remove and NaN values
kable(head(cortex_df_sub_t_scale[,1:6]),digits = 6) #check
```


The sum of all 'colSums' should also be zero
```{r check_scaling, warning = FALSE}
sum(colSums(cortex_df_sub_t_scale))
```

***Note:***  Very close! This is small enough
***Note:***  In fact, we could do the scaling during the PCA calculation! However, it is good for you to see what it actually does to the data

## 3. principal component analysis
It is to start with the real practical! Here, you will use PCA dimensionality reduction technique to reduce the data into 'meaningful' dimensions. For this, we will use hte prcomp() package 
Use the prcomp() package to perform the PCA. Save the result in pca_mod

***Information on prcomp()*** \
You can check <https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/prcomp> for additional information

### 3.1 calculate the PCA
First, we will perform the PCA using the features (genes) as dimensions, where each cell is an independent measurement
```{r pca calculate}
pca_adult_cortex <- prcomp(cortex_df_sub_t_scale) # If using the raw data martix, we can use ", scale = TRUE?
#summary(pca_adult_cortex)
```

### 3.2 Visualise the variation explained by PCs
We can have a look at the variation explained by each component using fviz_eig() function of the factoextra package

```{r checl_eigen_factors}
fviz_eig(pca_adult_cortex,
         barfill = 'purple',
         addlabels = TRUE,
         main = 'PCA analysis of the mouse cortex scRNAseq data')
```

##### ***Question 2***
**2. What does percentage of explained variance mean?** \
$~$  
**TIP: Dont forget, we have caluclated more dimensions** \
$~$  

### 3.3 Check top contributors to each PC
We can then check which genes contribute to the first PC1 the most

```{r check_PC1_contribution}
loadings_pc1 <- pca_adult_cortex$rotation[, 1]
loadings_pc1[which.max(abs(loadings_pc1))] 
```


##### ***Practice 1***
**1. Can you find the top 5 genes that contribute to PC2?**
$~$  
**TIP: Modify the code above and use order() function**
$~$  


## 3.4 plot the PCA results
We can plot the results using ggplot. Instead of points, we will use the country names

```{r plot_pca_1}

gg_pca <- 
  rbind(pca_adult_cortex$x[,1:2]) %>% 
  as_tibble() %>% 
  mutate(name = rownames(cortex_df_sub_t), 
         type = rep("cells", length(rownames(cortex_df_sub_t))))

gg_pca %>%
  ggplot(aes(x = PC1, y = PC2, shape = type, colour = type)) + #, label = name
  geom_point() +
  #geom_text(hjust = 0, nudge_x = 0.05) +
  coord_fixed() +
  theme_minimal() +
  theme(legend.position = "none") +
  labs(x = "PC 1", y = "PC 2", 
       title = "PCA of scRNAseq of cortex")
```

##### ***Practice 2***
**Plot the differences between two other dimension and answer the question below**
$~$ 


##### ***Questions 2***
**What do you conclude from these plots?**
$~$  
**TIP: Think of population/groups of cells, or differences between PCs**
$~$


##### IF YOU HAVE TIME #####
$~$
**Repeat the whole analysis by random sampling a selected cells**
$~$

## **BREAK**

## 4. Seurat analysis
It is time for the scRNAseq analysis! We will use the Seurat object that we have uploaded. This object made specially for the Seurat pipeline has a lot of 'slots' where information can be stored, and the architecture is a bit complicated. You do not need it for this tutorial, except what is mentioned

### 4.1 Downsize the dataset
To save from time, we will also subset the seurat object by selecting 1000 random cells. For this, we can use the subset() functuion

```{r Seurat_downsize}
adult_cortex_small <- subset(adult_cortex, downsample = 1000)
dataset = adult_cortex_small #rename for convenience
dataset
```


### 4.2 Metadata
This is where 'cell level' information is stored. This means there is one value for each cell
```{r metadata}
kable(head(dataset@meta.data[,1:6]),digits = 6)
```


### 4.3 Plot some quality metrics
An important metric is the number of RNA molecules (nCount_RNA) and genes (nFeature_RNA) per cell. These are automatically calculated when the Seurat object is generted form a data matrix.

```{r Seurat_QC_plot}
VlnPlot(object = dataset, features = c("nCount_RNA", "nFeature_RNA"),
				pt.size = .01)
```
### Standard pre-processing workflow ###

### 4.4 Calculate additional QC metrics
Start by generating  QC metrics additional to the no of genes/features

**Mitochondrial** RNA is the mRNA that is generated by the mitochondrial genome. Normally, these consitutte a small fraction of the total mRN. However, in dying or damaged cells, while cytoplasmic/nuclear mRNA degrades rapidly, mitochondrial mRNA is rather well preserved. Thus, a high ratio of mitochondrial mRNA indicates BAD cell quality 

mRNA coding for the **Ribosomal** subunit proteins is abundant (not to be confused with rRNA). Usually, a high ribosomal RNA percentage indicates production of a lot of proteins, and is very high in dividing cells or some secretory cells that need to constantly produce proteins. However, if most of the mRNA that we detect is ribosomal, it means that we wont learn much from this cell, and that we should exclude it from the analysis.

```{r Seurat_calculate_ribo_mito}
dataset <- PercentageFeatureSet(dataset,pattern='^mt-', col.name = "percent.mt")
dataset <- PercentageFeatureSet(dataset,pattern='Rp(s|l)', col.name = "percent.ribo")
```

### 4.5 Plot QC metrics
We can use the VlnPlot() function of the Seurat package to visualise the QC metrics

```{r Seurat_plot_ribo_mito, fig.width = 10, fig.asp = .5}
plot0 <- VlnPlot(object = dataset, features = c("percent.mt", "percent.ribo"),pt.size = 0, cols = "blue")
plot0
```

Let's visualize how mito and ribo percentages change as a function of the number of counts

```{r plot_scatter, fig.width = 8,fig.asp = .8}
plot1 <- FeatureScatter(dataset, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(dataset, feature1 = "nCount_RNA", feature2 = "percent.ribo")
plot3 <- FeatureScatter(dataset, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot_null <- ggplot() + theme_void()
(plot1 + plot2) / (plot3 + plot_null)
```


##### ***Questions 2.1***
**What is the relationship between total number of RNA per cell (nCounts) and** \
  **i) mitochondrial RNA percentage?** \
  **ii) ribosomal RNA percentage?** \
  **ii) number of features?**
$~$  


### 4.6 Filter low quality cells

##### ***Practice 2.2***
**We do not want to have low quality cells in our data. Looking at the plot, determine which cells to get rid of**
$~$ 
Use subset() to fetch the  cells that fit in your description
```{r Seurat_filter_practice, eval =FALSE, include=TRUE}
cutoff_mito = 
cutoff_ribo =
dataset <- subset(x = dataset, subset = percent.mt < cutoff_mito & percent.ribo < cutoff_ribo)
```


### 4.7 Normalise
In Seurat, standard preprocessing workflow is replaced by a single command. However, it is good to see this part to learn each step. First, we will normalize the data. This is to get rid of the differences in total RNA counts between cells. In other words, we will equalize the total count number in each cell to a fixed number (e.g. 10000 RNA molecules per  cell)

```{r Seurat_normalise}
dataset <- NormalizeData(object = dataset, normalization.method = "LogNormalize", scale.factor = 10000)
```

### 4.8 Detection of variable genes across the single cells
We want to find out 'informative genes' that will explain biological differences to use in some of hte downstream applications. If a gene is expressed everywhere, that doesnt tell us much. However, if a gene is expressed in a subset of cells, this will cause 'variation'

We can detect these genes using FindVariableFeatures() function

```{r Seurat_FindVariableFeatures}
## Here, we select the top 1,000 highly variable genes (Hvg) for downstream analysis.
dataset <- FindVariableFeatures(object = dataset, selection.method = 'mean.var.plot', mean.cutoff = c(0.0125, 3), dispersion.cutoff = c(0.5, Inf))
length(x = VariableFeatures(object = dataset)) #3084

```

Identify the 10 most highly variable genes
```{r Seurat_top10}
top10 <- head(VariableFeatures(dataset), 10)
top10
```

Now visualise

```{r Seurat_plot_variablefeatures, fig.width = 8,fig.asp = .5}
## Plot
# plot variable features with and without labels
plot1 <- VariableFeaturePlot(dataset)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
```
***Note:Dispersion indicates variation, while red color shows 'significantly variable' genes***

# Now take the top 1000 highly variable genes and 
```{r Seurat_hvgenes}
hv.genes <- head(rownames(HVFInfo(object = dataset)), 1000)
head(hv.genes,n=100) # list the first 100
```


### 4.8 Scale the data and get rid of the confounders
We discussed scaling (standardisation) in the previous section.Here, we will only take the hv.genes to scale and use in downstream dimensionality reduction. 

We can also get rid of the confounding factors at this point. These factors introduce 'technical noise' to the data. For instance, the number of reads per cell can influence the amount of information in a cell and make it seem different from another cell with low RNA levels, even though they are similar cells. 

We will use the ScaleData() function of hte Seurat package. The confounding factors can be discarded using 'vars.to.regress'

```{r Seurat_scale}
dataset <- ScaleData(object = dataset, features = hv.genes, vars.to.regress = c("percent.mt","nFeature_RNA"))

```
###Dimentionality reduction###

### 4.8 PCA analysis
Performing Dimensionality reduction in Seurat has been made very simple! We can first calculate the PCA
```{r Seurat_PCA_calculate}
dataset <- RunPCA(object = dataset, features = hv.genes, verbose = FALSE,npcs = 50)
```

Plot PCA, and color by the number of counts
```{r Seurat_PCA_plot}
plot1 <- DimPlot(object = dataset, reduction = 'pca',dims = c(1,2))
plot1
```
It is also easy to find out genes that contribute to the + and - direction of the top PCs! We can also plot the level of contribution

```{r Seurat_PCA_summary}
print(dataset[["pca"]], dims = 1:5, nfeatures = 5) # First box
VizDimLoadings(dataset, dims = 1:2, reduction = "pca") # Second box
```
##### ***Questions 2.3***
**Some genes contribute more to the variation then others... What do these genes and principle components tell us?** \
$~$  


### 4.8 SNN Graph Construction
We will use a graph-based clustering algorithm discussed at the lecture. For this, we need to build a neighborhood graph. In this network graph, each cell will be a node, and their similarity in high dimensional space will become their edges

We could say that cells closest to each other reside in a neighborhood

```{r Seurat_fundNeughbors}
dataset <- FindNeighbors(object = dataset, dims = 1:20) 
```

### 4.9 Find clusters using the louvain algorithm
```{r Seurat_cluster_plot}
dataset <- FindClusters(object = dataset, resolution = 0.6) # changing the resolution will change the size/number of clusters! c(0.6, 0.8, 1)
VlnPlot(object = dataset, features = c("nCount_RNA", "nFeature_RNA"))
```


### 4.10 UMAP and t-SNE analysis
We will use the top Pcs to calculate the umap and tSNE coordinates
```{r Seurat_umap_tsne_calculate}
dataset <- RunUMAP(object = dataset, reduction = "pca", dims = 1:20, verbose = FALSE)
dataset <- RunTSNE(object = dataset, dims = 1:20, verbose = FALSE)
```

##### ***Practice 2.4***
**Do you want to see them? adjust the DimPlot function to show the PCA, UMAP and t-SNE results!**
$~$  
**Include the cluster colors to visualize clusters**
$~$ 

##### ***Questions 2.5***
**How do the clusters distribute on different plots?**
$~$  
**Why is there a differnce?**
$~$  


